File: recipes/claude/12-stream-thinking-blocks-to-ui/typescript/stream_thinking.ts
/**
 * Stream Thinking Blocks to User Interfaces in TypeScript.
 *
 * Demonstrates real-time Server-Sent Event (SSE) parsing, handling:
 * - content_block_start (type: "thinking")
 * - content_block_delta (type: "thinking_delta")
 * - content_block_delta (type: "signature_delta")
 * - content_block_stop
 * - content_block_start (type: "text")
 * - content_block_delta (type: "text_delta")
 * - content_block_stop
 *
 * Implements a state machine suitable for driving collapsible UI accordion components.
 */

import Anthropic from '@anthropic-ai/sdk';
import * as dotenv from 'dotenv';

dotenv.config();

const apiKey = process.env.ANTHROPIC_API_KEY;
const model = process.env.ANTHROPIC_MODEL || 'claude-3-7-sonnet-20250219';

if (!apiKey) {
  console.error('ERROR: ANTHROPIC_API_KEY environment variable is missing.');
  process.exit(1);
}

const anthropic = new Anthropic({ apiKey });

export type StreamUIState = 'idle' | 'thinking' | 'answering' | 'completed' | 'error';

export interface UIThinkingAccordionState {
  uiState: StreamUIState;
  thinkingContent: string;
  signature: string;
  textContent: string;
  thinkingStartTime: number | null;
  thinkingDurationMs: number;
  textStartTime: number | null;
  textDurationMs: number;
  isAccordionExpanded: boolean;
}

/**
 * Initializes fresh UI stream state.
 */
export function createInitialUIState(): UIThinkingAccordionState {
  return {
    uiState: 'idle',
    thinkingContent: '',
    signature: '',
    textContent: '',
    thinkingStartTime: null,
    thinkingDurationMs: 0,
    textStartTime: null,
    textDurationMs: 0,
    isAccordionExpanded: true,
  };
}

/**
 * Processes stream events and updates the UI thinking accordion state.
 */
export async function streamThinkingToUI(
  userPrompt: string,
  thinkingBudget: number = 2048,
  onStateUpdate?: (state: UIThinkingAccordionState) => void
): Promise<UIThinkingAccordionState> {
  const state = createInitialUIState();

  console.log(`[INIT] Streaming from ${model} with thinking budget: ${thinkingBudget} tokens`);
  console.log('='.repeat(65));

  const stream = await anthropic.messages.create({
    model,
    max_tokens: 4096,
    thinking: {
      type: 'enabled',
      budget_tokens: thinkingBudget,
    },
    stream: true,
    messages: [
      {
        role: 'user',
        content: userPrompt,
      },
    ],
  });

  for await (const event of stream) {
    switch (event.type) {
      case 'content_block_start': {
        const block = event.content_block;
        if (block.type === 'thinking') {
          state.uiState = 'thinking';
          state.thinkingStartTime = Date.now();
          state.isAccordionExpanded = true;
          console.log('\n>>> [UI EVENT] Thinking Block Started -> Accordion Opened');
        } else if (block.type === 'text') {
          state.uiState = 'answering';
          state.textStartTime = Date.now();
          // In web clients, collapse accordion automatically or keep minimized badge
          state.isAccordionExpanded = false;
          console.log('\n>>> [UI EVENT] Text Answer Started -> Accordion Collapsed');
        }
        onStateUpdate?.(state);
        break;
      }

      case 'content_block_delta': {
        const delta = event.delta;
        if (delta.type === 'thinking_delta') {
          state.thinkingContent += delta.thinking;
          process.stdout.write(`\r[THINKING ACCORDION] Streaming tokens... (${state.thinkingContent.length} chars)`);
          onStateUpdate?.(state);
        } else if (delta.type === 'signature_delta') {
          state.signature += delta.signature;
          console.log(`\n[SIGNATURE] Encrypted block signature captured (${state.signature.length} bytes)`);
          onStateUpdate?.(state);
        } else if (delta.type === 'text_delta') {
          state.textContent += delta.text;
          process.stdout.write(delta.text);
          onStateUpdate?.(state);
        }
        break;
      }

      case 'content_block_stop': {
        if (state.uiState === 'thinking') {
          if (state.thinkingStartTime) {
            state.thinkingDurationMs = Date.now() - state.thinkingStartTime;
          }
          console.log(`\n>>> [UI EVENT] Thinking Block Closed in ${state.thinkingDurationMs}ms`);
        } else if (state.uiState === 'answering') {
          if (state.textStartTime) {
            state.textDurationMs = Date.now() - state.textStartTime;
          }
          console.log(`\n>>> [UI EVENT] Text Block Closed in ${state.textDurationMs}ms`);
        }
        onStateUpdate?.(state);
        break;
      }

      case 'message_stop': {
        state.uiState = 'completed';
        console.log('\n>>> [UI EVENT] Stream Finalized');
        onStateUpdate?.(state);
        break;
      }
    }
  }

  console.log('\n' + '='.repeat(65));
  console.log('CLIENT ACCORDION METRICS:');
  console.log(`Thinking Duration: ${state.thinkingDurationMs} ms`);
  console.log(`Answer Duration:   ${state.textDurationMs} ms`);
  console.log(`Thinking Bytes:    ${state.thinkingContent.length}`);
  console.log(`Answer Bytes:      ${state.textContent.length}`);
  console.log(`Signature Length:  ${state.signature.length}`);
  console.log('='.repeat(65));

  return state;
}

// Direct execution entrypoint
async function main() {
  const prompt =
    'Design a real-time reactive event architecture in TypeScript for handling concurrent SSE streams without memory leaks. Provide the structural trade-offs first, followed by code.';

  try {
    await streamThinkingToUI(prompt, 2048);
  } catch (err) {
    console.error('Streaming failure:', err);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

File: recipes/claude/12-stream-thinking-blocks-to-ui/python/stream_thinking.py
"""
Async Streaming Thinking Blocks Parser for Anthropic Claude.

Demonstrates real-time parsing of Server-Sent Events (SSE) including:
- content_block_start (type: "thinking")
- content_block_delta (type: "thinking_delta")
- content_block_delta (type: "signature_delta")
- content_block_stop
- content_block_start (type: "text")
- content_block_delta (type: "text_delta")
- content_block_stop
- message_delta (token usage)
"""

from __future__ import annotations

import asyncio
import os
import sys
import time
from typing import Optional
from dotenv import load_dotenv

import anthropic

load_dotenv()

API_KEY = os.environ.get("ANTHROPIC_API_KEY")
MODEL = os.environ.get("ANTHROPIC_MODEL", "claude-3-7-sonnet-20250219")


class ThinkingStreamState:
    """Tracks state transitions across thinking and answer blocks."""

    def __init__(self) -> None:
        self.thinking_text: str = ""
        self.signature: str = ""
        self.answer_text: str = ""
        self.current_block_type: Optional[str] = None
        self.thinking_started_at: Optional[float] = None
        self.thinking_elapsed_ms: float = 0.0
        self.text_started_at: Optional[float] = None
        self.text_elapsed_ms: float = 0.0
        self.output_tokens: int = 0


async def stream_thinking_response(prompt: str, budget_tokens: int = 2048) -> ThinkingStreamState:
    """
    Initiates an asynchronous stream with extended thinking enabled,
    routing thinking deltas and text deltas to dedicated handlers.
    """
    client = anthropic.AsyncAnthropic(api_key=API_KEY)
    state = ThinkingStreamState()

    print(f"Connecting to {MODEL} with thinking budget of {budget_tokens} tokens...")
    print("-" * 65)

    async with client.messages.stream(
        model=MODEL,
        max_tokens=4096,
        thinking={
            "type": "enabled",
            "budget_tokens": budget_tokens,
        },
        messages=[
            {
                "role": "user",
                "content": prompt,
            }
        ],
    ) as stream:
        async for event in stream:
            event_type = event.type

            if event_type == "content_block_start":
                block = event.content_block
                state.current_block_type = block.type

                if block.type == "thinking":
                    state.thinking_started_at = time.perf_counter()
                    print("\n[STATE: THINKING STARTED] Collapsible accordion expanding...")
                elif block.type == "text":
                    state.text_started_at = time.perf_counter()
                    print("\n[STATE: ANSWER STARTED] Thinking complete. Rendering answer...")

            elif event_type == "content_block_delta":
                delta = event.delta

                if delta.type == "thinking_delta":
                    thinking_chunk = delta.thinking
                    state.thinking_text += thinking_chunk
                    # Display stream indicator or thinking snippet
                    sys.stdout.write(f"\r[THINKING ACCORDION] Buffered {len(state.thinking_text)} characters...")
                    sys.stdout.flush()

                elif delta.type == "signature_delta":
                    state.signature += delta.signature
                    print(f"\n[SIGNATURE] Captured verification signature ({len(state.signature)} chars)")

                elif delta.type == "text_delta":
                    text_chunk = delta.text
                    state.answer_text += text_chunk
                    sys.stdout.write(text_chunk)
                    sys.stdout.flush()

            elif event_type == "content_block_stop":
                if state.current_block_type == "thinking":
                    if state.thinking_started_at:
                        state.thinking_elapsed_ms = (time.perf_counter() - state.thinking_started_at) * 1000
                    print(f"\n[STATE: THINKING CLOSED] Finalized in {state.thinking_elapsed_ms:.1f}ms")
                elif state.current_block_type == "text":
                    if state.text_started_at:
                        state.text_elapsed_ms = (time.perf_counter() - state.text_started_at) * 1000
                    print(f"\n[STATE: ANSWER COMPLETE] Generated in {state.text_elapsed_ms:.1f}ms")

                state.current_block_type = None

            elif event_type == "message_delta":
                if hasattr(event, "usage") and event.usage:
                    state.output_tokens = event.usage.output_tokens

    print("\n" + "=" * 65)
    print("STREAMING METRICS SUMMARY:")
    print(f"Thinking Duration: {state.thinking_elapsed_ms:.2f} ms")
    print(f"Answer Duration:   {state.text_elapsed_ms:.2f} ms")
    print(f"Thinking Length:   {len(state.thinking_text)} characters")
    print(f"Answer Length:     {len(state.answer_text)} characters")
    print(f"Total Output Tokens: {state.output_tokens}")
    print("=" * 65)

    return state


async def main() -> None:
    if not API_KEY:
        print("ERROR: ANTHROPIC_API_KEY is not configured.", file=sys.stderr)
        print("Set it via: export ANTHROPIC_API_KEY='sk-ant-api03...'", file=sys.stderr)
        sys.exit(1)

    prompt = (
        "Develop an algorithmic strategy to calculate the median of two sorted "
        "arrays of different sizes in O(log(min(n, m))) time complexity. "
        "Detail your internal trade-offs before presenting the final Python implementation."
    )
    await stream_thinking_response(prompt, budget_tokens=2048)


if __name__ == "__main__":
    asyncio.run(main())
